crates/revm/src/evm_impl.rs

```diff
@@ -143,6 +181,42 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> Transact<DB::Error>
 
         // load acc
         let journal = &mut self.data.journaled_state;
+
+        #[cfg(feature = "optimism")]
+        if self.data.env.cfg.optimism {
+            // If the transaction is a deposit with a `Some` `mint` value, add the minted value
+            // in wei to the caller's balance. This should be persisted to the database prior
+            // to the rest of execution below.
+            if let Some(mint) = tx_mint {
+                journal
+                    .load_account(tx_caller, self.data.db)
+                    .map_err(EVMError::Database)?
+                    .0
+                    .info
+                    .balance += U256::from(mint);
+                journal.checkpoint();
+            }
+
+            // If the transaction is not a deposit transaction, subtract the L1 data fee from the
+            // caller's balance directly after minting the requested amount of ETH.
+            #[cfg(feature = "optimism")]
+            if let Ok(l1_block_info) = optimism::fetch_l1_block_info(&mut self.data.db) {
+                if !is_deposit {
```

Comment:
perf: change the nesting order of these conditionals, no need to read the DB if the tx isn't a deposit.